### DISCUSSION TITLE
Access params in server components which are not page.tsx or layout.tsx

### DISCUSSION THREAD

**Discussion by RowinVanAmsterdam** [vercel/next.js#63298] [App Router] [answered]
Currently I'm working on a multi-language webapp, build in NextJS 14 (with app router) and I want to share the locale between server components.

In client component it's easy, because I can use the `useParams` hook in order to retrieve the locale from the URL. However in server components I don't see a way to handle that.

In your `page.tsx` or `layout.tsx` file (both server components) you can simply do this:
```
type HomeProps = {
  params: {
    locale: string;
  };
};

export default async function Home(props: HomeProps) {
 const { params } = props;
 console.log(params.locale) // --> results in en-US for example if url = .com/en-US

```

but if I have a server component, let's say `Hero.tsx`, I can't do that or I should pass it as a prop, like so: 

`<Hero locale={params.locale } />`

Of course I can pass the props from page.tsx to Hero.tsx, but this can result in prop drilling and I would prefer avoiding that.

**Comment by mutasim77** [comment]
Yup, that's correct! But there's a way to do so using middleware. Check out this [discussion](https://github.com/vercel/next.js/issues/43704#issuecomment-1411186664)

**Reply by RowinVanAmsterdam** [reply]
Thanks for your response. I just checked it out, but it seems like the workaround disables SSG completely.

**Comment by icyJoseph** [comment, marked as answer]
Hi,

There's one cool way to do this. It uses `cache` from React.

The insight here is that `cache`'s memoization is scoped to the current request. 

- https://react.dev/reference/react/cache#caveats

So what you can do is, have a function that returns a container object, say a Map, or a plain JavaScript object, or a Set, etc...

Pass that function into cache (it can be inlined too). This gives you back a function that's tracked by React.

Now, when you invoke the function, React saves the result, and any further invocations receive the same container object. Which means that you can set/get data from the same container, during the ongoing request, from various parts of the React server tree.

That works essentially as a shared store during the SSR pass. 

Here's a demo: https://stackblitz.com/edit/nextjs-q3dnp8?file=app%2Fpage.tsx,app%2FLocaleSwitch.tsx,app%2F[locale]%2Fpage.tsx,app%2Fserver-context.ts


```tsx
import 'server-only';
import { cache } from 'react';

const requestContext = cache(() => {
  return new Map<string, string>();
});

export const setRequestContext = (key: string, value: string) =>
  requestContext().set(key, value);

export const getRequestContext = (key: string) => requestContext().get(key);
```

This is but one way to do it, there's also a library that implements this for you: https://www.npmjs.com/package/server-only-context but I think it is a good exercise to understand how it really works.

---

I still believe that people are taking dislike for prop-drilling to a far extreme. Passing props, is a valid pattern, and it is one that makes the open API of a component, public and predictable. 

When a component starts to rely on Context, it is a form of dependency injection that makes things less modular, often shown during unit testing, or when trying to re-use components in other trees, since you necessarily have to make the entire tree dependent on the Context Provider.

That doesn't mean that Context Providers are also bad either. It is our job as devs to find the right balance, and know when teleporting data makes more sense, and when we want to enforce some sort of explicit contract through prop passing. 

Passing unused props several levels, is not ideal, just as it is not ideal, to rely on context to reach the very end of your component tree. (where in order to re-use a button component, you need to bring their context provider too). This is where the component spectrum comes into play, and knowing where to put boundaries to the type of data the next level of components get to see.

**Reply by RowinVanAmsterdam** [reply]
Appreciate your explanation and suggestions. I just tried it out and played with it a bit. It seems to work, as long as I set the context on a page.tsx or layout.tsx level. Once I tried setting it from the middleware, it's not possible and wondering if there are any other caveats besides this one and the ones you were mentioning. 

I do think prop drilling would be sufficient enough in my case and might also be the route. As you said it is not the end of the world. But it just feels a bit wrong to do it for each (server) component, just to have translations in it. 

If you're interested I receive the following error when I try to set it from the middleware:
```
src/utils/serverContext.ts (4:0) @ <unknown>
 ⨯ (0 , react__WEBPACK_IMPORTED_MODULE_1__.cache) is not a function
  2 | import { cache } from 'react';
  3 |
> 4 | const requestContext = cache(() => {
  5 |   return new Map<string, string>();
  6 | });
  7 |
```

**Reply by icyJoseph** [reply]
Middleware might as well be an entirely different system. It doesn't share any JavaScript context whatsoever with your rendering pass.

**Reply by icyJoseph** [reply]
And you should set your globals, preferably at page level. Pages render before layout.

**Reply by RowinVanAmsterdam** [reply]
That makes sense and also explains the current behaviour. But I'm wondering at what level you're "setting" the `setRequestContext` then. Do you do it on the highest/start page.tsx? Since page is running before the layout. Also you don't want to set it in multiple page.tsx files, because once should be enough if I'm not mistaken. 

Also did some additional reading, but if I understand it correctly there is no other way to do this. Libraries that are providing a similar functionality all uses this cache method currently. 

Read some other solution as well, but they definitely feel more like "hacking". Like doing a fetch request, getting it from headers etc.. All with their own caveats. 

--

Edit: another method I just found is using/reading it from your function [cookies](https://nextjs.org/docs/app/api-reference/functions/cookies). For example:

`request.cookies.get('NEXT_LOCALE')`

which returns object: `{ name: 'NEXT_LOCALE', value: 'en-US' }` 

--> Nvm, using this method would expects passing down NextRequest as a prop.

**Reply by icyJoseph** [reply]
You should set it as soon as you have access to the data you need.

Since it is a map, you can read and set as you like. But typically you want the parent components to do that, so that the children can read it.

Well, that "hack" opinion is starting to fall apart. The React cache has matured, and it's purpose is to memoize:

> cache lets you cache the result of a data fetch or computation - https://react.dev/reference/react/cache#noun-labs-1201738-(2)

I can understand the annoyance that this is a canary feature, and a new thing to learn, but why we refuse to acknowledge the `or computation` bit, I don't know.

Why is it not a hack when the cache holds onto a promise, but it is a hack when it holds onto a Map 🤷‍♂️

It reminds me of AsyncLocalStorage in Node.js, which takes in a store and a callback that triggers asynchronous work. That asynchronous work, will have access to the store provided, via the AsyncLocalStorage API, so that you can have several asynchronous workloads, all accessing a potentially unique store.

One use case for that was correlation ids to debug issues in your server logs, but you could also store there data that the request needs "globally".

The cache is a different flavor of that, probably somewhat similar, I haven't looked at the details of implementation. Though since the rendering (function calling) is done by React, you can't really do your own AsyncLocalStorage solution. And that's where cache kicks in.

**Reply by RowinVanAmsterdam** [reply]
Hmm you're right, this also seems like the only plausible way now to do handle this. I think I couldn't describe it better than you already do. It may be several factors why it isn't fully accepted yet, or better say refused by many. Like you already say it's something new you have to understand and perhaps a bit daunting if not being mentioned in any official docs to use it like this. 

But I can be wrong of course. It just feels new and that is always a bit scary, especially when it can be something as fundamental for a (large) project. But I really appreciate this discussion and your explanation. I played with it a bit already and I'm definitely going to use it.
